File: app/site_settings.py
"""Vault-wide settings and validation."""

import datetime as dt

from sqlalchemy import select
from sqlalchemy.orm import Session

from .models import VaultSetting

SITE_SETTING_DEFAULTS: dict[str, object] = {
    "archivePermanentDeleteAdminOnly": True,
}
BOOLEAN_SITE_SETTINGS = {"archivePermanentDeleteAdminOnly"}


def normalize_site_settings(raw: object) -> dict[str, object]:
    normalized = dict(SITE_SETTING_DEFAULTS)
    if not isinstance(raw, dict):
        return normalized
    for key in BOOLEAN_SITE_SETTINGS:
        value = raw.get(key)
        if isinstance(value, bool):
            normalized[key] = value
    return normalized


def clean_site_setting_patch(raw: object) -> dict[str, object]:
    if not isinstance(raw, dict):
        raise ValueError("Settings must be an object")
    cleaned: dict[str, object] = {}
    for key, value in raw.items():
        if key not in SITE_SETTING_DEFAULTS:
            raise ValueError(f"Unknown setting: {key}")
        if key in BOOLEAN_SITE_SETTINGS:
            if not isinstance(value, bool):
                raise ValueError(f"{key} must be a boolean")
            cleaned[key] = value
    return cleaned


def site_settings_for_db(db: Session) -> dict[str, object]:
    rows = db.execute(select(VaultSetting)).scalars().all()
    return normalize_site_settings({row.key: row.value for row in rows})


def merge_site_settings(db: Session, patch: dict[str, object]) -> dict[str, object]:
    cleaned = clean_site_setting_patch(patch)
    existing = {row.key: row for row in db.execute(select(VaultSetting)).scalars().all()}
    timestamp = dt.datetime.now(tz=dt.UTC)
    for key, value in cleaned.items():
        row = existing.get(key)
        if row:
            row.value = value
            row.updated_at = timestamp
        else:
            db.add(VaultSetting(key=key, value=value, updated_at=timestamp))
    db.flush()
    return site_settings_for_db(db)


def archive_permanent_delete_admin_only(db: Session) -> bool:
    return bool(site_settings_for_db(db)["archivePermanentDeleteAdminOnly"])

File: crates/vault-server/src/site_settings.rs
use serde_json::{Map, Value, json};
use sqlx::SqlitePool;
use thiserror::Error;

use crate::state_events::state_event_resources_json;

#[derive(Debug, Error)]
pub enum SiteSettingsError {
    #[error("{0}")]
    InvalidPatch(String),
    #[error(transparent)]
    Database(#[from] sqlx::Error),
    #[error(transparent)]
    Json(#[from] serde_json::Error),
}

pub async fn site_settings_for_db(pool: &SqlitePool) -> Result<Value, SiteSettingsError> {
    let rows = sqlx::query_as::<_, (String, String)>("SELECT key, value FROM vault_settings")
        .fetch_all(pool)
        .await?;
    let mut raw = Map::new();
    for (key, value) in rows {
        raw.insert(key, serde_json::from_str::<Value>(&value)?);
    }
    Ok(normalize_site_settings(&Value::Object(raw)))
}

pub async fn archive_permanent_delete_admin_only(
    pool: &SqlitePool,
) -> Result<bool, SiteSettingsError> {
    Ok(site_settings_for_db(pool)
        .await?
        .get("archivePermanentDeleteAdminOnly")
        .and_then(Value::as_bool)
        .unwrap_or(true))
}

pub async fn update_admin_site_settings(
    pool: &SqlitePool,
    raw_patch: &Value,
) -> Result<Value, SiteSettingsError> {
    let patch = clean_site_setting_patch(raw_patch)?;
    let mut transaction = pool.begin().await?;
    for (key, value) in patch {
        sqlx::query(
            r"
            INSERT INTO vault_settings (key, value, updated_at)
            VALUES (?, ?, CURRENT_TIMESTAMP)
            ON CONFLICT(key)
            DO UPDATE SET
                value = excluded.value,
                updated_at = CURRENT_TIMESTAMP
            ",
        )
        .bind(key)
        .bind(value.to_string())
        .execute(&mut *transaction)
        .await?;
    }
    sqlx::query(
        r"
        INSERT INTO state_events (event_type, resources)
        VALUES ('admin.settings.updated', ?)
        ",
    )
    .bind(state_event_resources_json(&["admin", "settings"]))
    .execute(&mut *transaction)
    .await?;
    transaction.commit().await?;
    site_settings_for_db(pool).await
}

#[must_use]
pub fn normalize_site_settings(raw: &Value) -> Value {
    let mut normalized =
        Map::from_iter([("archivePermanentDeleteAdminOnly".to_string(), json!(true))]);
    if let Some(value) = raw
        .as_object()
        .and_then(|object| object.get("archivePermanentDeleteAdminOnly"))
        .and_then(Value::as_bool)
    {
        normalized.insert("archivePermanentDeleteAdminOnly".to_string(), json!(value));
    }
    Value::Object(normalized)
}

pub fn clean_site_setting_patch(raw: &Value) -> Result<Map<String, Value>, SiteSettingsError> {
    let Some(raw_object) = raw.as_object() else {
        return Err(invalid_patch("Settings must be an object"));
    };
    let mut cleaned = Map::new();
    for (key, value) in raw_object {
        match key.as_str() {
            "archivePermanentDeleteAdminOnly" => {
                let Some(value) = value.as_bool() else {
                    return Err(invalid_patch(format!("{key} must be a boolean")));
                };
                cleaned.insert(key.clone(), json!(value));
            }
            _ => return Err(invalid_patch(format!("Unknown setting: {key}"))),
        }
    }
    Ok(cleaned)
}

fn invalid_patch(message: impl Into<String>) -> SiteSettingsError {
    SiteSettingsError::InvalidPatch(message.into())
}
